public/friendlycode/js/fc/publish.js

```diff
@@ -0,0 +1,172 @@
+      unpublishBtn.children("span").text("Unpublish");
+    });
+  }
+
+  function saveAll() {
+    // TODO: Logic to save all unsaved files in bramble
+    // Blocked on https://github.com/humphd/brackets/issues/372
+  }
+
+  function publish() {
+    saveAll();
+    publishBtn.off("click", publish);
+    publishBtn.text("Publishing...");
+    updatePublishBtn.off("click", publish);
+    updatePublishBtn.text("Publishing...");
+
+    var project = {
+      description: projectDesc.val() || " ",
+      public: isPublic,
+      dateUpdated: (new Date()).toISOString()
+    };
+
+    var request = $.ajax({
+      contentType: "application/json",
+      headers: {
+        "X-Csrf-Token": csrfToken,
+        "Accept": "application/json"
+      },
+      type: "PUT",
+      url: url + "/publish",
+      data: JSON.stringify(project)
+    });
+    request.done(function(project) {
+      if(request.status !== 200) {
+        console.error("[Bramble] Server was unable to publish project, responded with status ", request.status);
+        return;
+      }
+
+      $("#publish-link > a")
+        .attr("href", project.link)
+        .text(project.link);
+      $("#publish-changes").addClass("hide");
+      $("#publish-live").removeClass("hide");
+      unpublishBtn.off("click", unpublish).on("click", unpublish);
+    });
+    request.fail(function(jqXHR, status, err) {
+      console.error("[Bramble] Failed to send request to publish project to the server with: ", err);
+    });
+    request.always(function() {
+      publishBtn.on("click", publish);
+      publishBtn.text("Publish");
+      updatePublishBtn.on("click", publish);
+      updatePublishBtn.text("Update published version");
+    });
+  }
+
+  function init(editor, options) {
+    bramble = editor;
+    var project = options.project || {};
+    var fileSync = options.sync;
+    url = options.appUrl;
+    csrfToken = $("meta[name='csrf-token']").attr("content");
+    publishBtn = $("#publish-button-publish");
+    updatePublishBtn = $("#publish-button-update");
+    unpublishBtn = $("#publish-button-unpublish");
+    projectDesc = $("#publish-details > textarea.publish-description");
+    projectPublic = $("#publish-public-gallery-toggle>ellipse");
+    publishBtn.on("click", publish);
+
+    $("#publish-public-gallery-toggle").click(function() {
+      isPublic = !isPublic;
+      setPublicGalleryToggle();
+    });
+
+    if(project && project.description) {
+      projectDesc.val(project.description);
+    }
+
+    if(project.publishUrl) {
+      $("#publish-link > a")
+        .attr("href", project.publishUrl)
+        .text(project.publishUrl);
+      $("#publish-changes").addClass("hide");
+      $("#publish-live").removeClass("hide");
+      unpublishBtn.on("click", unpublish);
+    }
+
+    if(fileSync) {
+      fileSync.setCallback(function() {
+        $("#publish-changes").removeClass("hide");
+        updatePublishBtn.off("click", publish).on("click", publish);
```

Comment:
Nit: I know why you're doing it, but this on/off stuff is kinda gross. Is there a way to check if the event is registered? It would make this stuff easier to read.